### ISSUE TITLE
Should call jedisPool.destroy

### ISSUE THREAD

**Issue by k4200** [cb372/scalacache#45]
Hi,

## Summary

I'm having a problem that a program runs out of connections to Redis, and guessing that it might be caused by scalacache-redis.

I'm not 100% sure, but I think one of the three options below needs to be taken:
1. Use `jedisPool.returnResource(jedis)` instead of `jedis.close()`
1. Make `RedisCache#jedisPool` public, so client code can call `JedisPool#destroy`
1. Instantiate plain `Jedis` every time instead of getting one from the pool

### jedis.close is a problem?
Here is a sample code from the Jedis [wiki](https://github.com/xetorthio/jedis/wiki/Getting-started):
```scala
Jedis jedis = null;
try {
  jedis = pool.getResource();
  /// ... do stuff here ... for example
  jedis.set("foo", "bar");
  String foobar = jedis.get("foo");
  jedis.zadd("sose", 0, "car"); jedis.zadd("sose", 0, "bike"); 
  Set<String> sose = jedis.zrange("sose", 0, -1);
} finally {
  if (jedis != null) {
    jedis.close();
  }
}
/// ... when closing your application:
pool.destroy();
```

I also took a brief look at Jedis source code, and it seems `Jedis#close` actually closes the connection. The sample code above calls `pool.destroy()` at the end, so it doesn't matter, but I don't think client code should close a connection taken from the connection pool (in general, no?)

This is an old thread, so I'm not sure this is still valid now,
https://groups.google.com/forum/?fromgroups=#!topic/jedis_redis/UeOhezUd1tQ

but it says:
> Returning the resource to the Pool is important, so remember to do it. Otherwise when closing your app it'll wait for the resource to return.

The sample code and wiki page recommend two different ways, and they are kinda contradicting, but in any case, I guess calling only `Jedis#close` isn't enough.

Also, another Jedis wiki [page](https://github.com/xetorthio/jedis/wiki/AdvancedUsage) says the following:
> Forgetting pool.destroy keeps the connection open until timeout is reached.

### My conclusion

Now that we've got two options.

> Returning the resource to the Pool is important, so remember to do it. Otherwise when closing your app it'll wait for the resource to return.

If the above sentence is still valid, the option below is the way to go:
> Use `jedisPool.returnResource(jedis)` instead of `jedis.close()`

If we want to follow the sample doc, we want to call `JedisPool#destroy`, so the following is an option:
> Make `RedisCache#jedisPool` public, so client code can call `JedisPool#destroy`

We also have an option of not using connection pooling, and sticking to plain `Jedis`, in which case, `Jedis#close` should be a valid way when you finish using the connection.

Thanks.

EDIT:
Forgot to mention that [sedis](https://github.com/pk11/sedis/blob/master/src/main/scala/sedis.scala) uses `returnResourceObject`, which is a bit different from `returnResource`.

**Comment by cb372** [comment]
Thank you for the very thorough investigation! I'll take a proper look at this asap.

**Comment by cb372** [comment]
In the latest versions of Redis, `JedisPool.returnResource()` is deprecated and `Jedis.close()` is the recommended method to use:
https://github.com/xetorthio/jedis/blob/jedis-2.6.3/src/main/java/redis/clients/jedis/JedisPool.java#L102

Anyway, it shouldn't matter which one we call, because internally `Jedis.close()` calls `JedisPool.returnResource(jedis)`. When you borrow a `Jedis` using `JedisPool.getResource()`, it sets the `Jedis` object's `dataSource` field (https://github.com/xetorthio/jedis/blob/jedis-2.6.0/src/main/java/redis/clients/jedis/JedisPool.java#L100). Which means that when you call `Jedis.close()` it will not close the client, but instead return it to the pool (https://github.com/xetorthio/jedis/blob/jedis-2.6.0/src/main/java/redis/clients/jedis/Jedis.java#L3580)

So, as far as I know, scalacache-redis is using Jedis and JedisPool correctly, and it shouldn't be leaking.

You're right that we should make `RedisCache#jedisPool` a field so you can call `destroy()` on it. Or maybe add a `close()` method to `Cache` that takes care of this for you. But you only need to call this when you shutdown your app, so it is unrelated to any connection leak.

**Comment by cb372** [comment]
By the way, are you sure it's a connection leak? Or is your application simply trying to use too many Jedis connections at once? By default, the Jedis connection pool size is only 8. You could try increasing it, like so:

```
val poolConfig = new GenericPoolConfig
poolConfig.setMaxTotal(123)
val jedisPool = new JedisPool(poolConfig, host, port)
implicit val scalaCache = ScalaCache(RedisCache(jedisPool))
```

**Comment by k4200** [comment]
Thanks for the investigation.

> are you sure it's a connection leak?

Not so sure, actually. I use scalacache-redis for our web application and batch program. The other component that uses Redis is Play! framework Redis plugin. Below are the facts that I observe:

* The batch job runs periodically, but it waits for something at the end of it, and the process remains
* After a certain amount of time, both the web and the batch become unable to connect to Redis
```
play.api.Application$$anon$1: Execution exception[[JedisConnectionException: Could not get a resource from the pool]]
```
* Restarting the web didn't solve the issue, but restarting Redis did

My wild *guess* was if the batch could issue `jedisPool.destroy` at the end, it would finish normally, and therefore it would release the resources. Again, I did't find the root cause of the issue, so there might be other factors than scalacache-redis that caused it.

As for the pool size, the service is in beta and not many connections should be used, but I'll increase it for safety. Thanks for the info.

The reason why restarting the web didn't resolve the issue is maybe, the OS didn't detect disconnection and connections stayed "established". 

I did the following workarounds and `Could not get a resource from the pool` hasn't occurred since then, but the batch still waits for something and the process remains:
* Set `timeout 300` in redis.conf
* Lower the `net.ipv4.tcp_keepalive_*`

I'll look deeper into the issue, and get back to you. Thanks for your help.

**Comment by k4200** [comment]
I investigated the issue a bit deeper.

> * The batch job runs periodically, but it waits for something at the end of it, and the process remains
> * After a certain amount of time, both the web and the batch become unable to connect to Redis

It turned out that the former was caused by another component, and now it's been fixed. The latter can still happen theoretically, but it's mitigated by the workarounds mentioned above, so it's not really a problem now.

I'll report it again if I find new information / potential problems related to this.

Thanks again for your help.
